### PR TITLE
[google_maps_flutter] Google maps flutter marker clustering support

### PR DESCRIPTION
**Pull request by jokerttu**
Adds support for marker clustering for all platforms

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart [hunk 1]
```diff
@@ -1161,6 +1161,57 @@ void main() {
...
   );
+
+  testWidgets('marker clustering', (WidgetTester tester) async {
+    const int clusterManagersAmount = 2;
+    const int markersPerClusterManager = 25;
+    final Set<Marker> markers = <Marker>{};
+    final Set<ClusterManager> clusterManagers = <ClusterManager>{};
+
+    for (int i = 0; i < clusterManagersAmount; i++) {
+      const ClusterManagerId clusterManagerId = ClusterManagerId('cluster');
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart:1172]: Should you use that paramater i in the clusterID? Now both clusters have same id. Dunno if this is a problem, but sounds like somethings that should be avoided.

File: packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      // Remove cluster manager.
+      clusterManagers.remove(clusterManager.clusterManagerId);
+    });
+  }
+
+  void _addMarkersToCluster(ClusterManager clusterManager) {
+    for (int i = 0; i < 15; i++) {
+      final String markerIdVal =
+          '${clusterManager.clusterManagerId.value}_marker_id_$_markerIdCounter';
+      _markerIdCounter++;
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart:118]: Why? Not used in the function...

File: packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart (hunk 1 above)
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart:1172]: Yes I should, good catch... don't know why test passed :D need to check that as well.

File: packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart (hunk 2 above)
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart:118]: It's used one line above

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m [hunk 3]
```diff
@@ -83,13 +88,27 @@ - (void)setRotation:(CLLocationDegrees)rotation {
 }
 
 - (void)setVisible:(BOOL)visible {
-  self.marker.map = visible ? self.mapView : nil;
+  // If marker belongs the cluster manager, visibility need to be controlled with the opacity
+  // as the cluster manager controls when marker is on the map and when not.
+  if (self.clusterManagerId != (id)[NSNull null]) {
+    self.marker.opacity = visible ? self.marker.opacity : 0.0f;
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:94]: The marker is not visible but does it respond to click?

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m [hunk 4]
```diff
@@ -247,37 +275,67 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
...
+  NSString *clusterManagerId = markerToAdd[@"clusterManagerId"];
+  CLLocationCoordinate2D position = [FLTMarkersController getPosition:markerToAdd];
+  GMSMarker *marker = [GMSMarker markerWithPosition:position];
+  FLTGoogleMapMarkerController *controller =
+      [[FLTGoogleMapMarkerController alloc] initWithMarker:marker
+                                                identifier:identifier
+                                          clusterManagerId:clusterManagerId
+                                                   mapView:self.mapView];
+  [controller interpretMarkerOptions:markerToAdd registrar:self.registrar];
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:293]: Hmm, does that first check work correctly? Can clusterManagerId be nil? Is empty string allowed?

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m [hunk 5]
```diff
@@ -247,37 +275,67 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
...
+  }
+}
+
+- (void)removeMarker:(NSString *)identifier {
+  FLTGoogleMapMarkerController *controller = self.markerIdentifierToController[identifier];
+  if (!controller) {
+    return;
+  }
+  NSString *clusterManagerId = [controller clusterManagerId];
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:333]: Same check here as in line 293. Maybe create separate function for this check?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+/// Uniquely identifies a [ClusterManager] among [GoogleMap] clusters.
+///
+/// This does not have to be globally unique, only unique among the list.
+@immutable
+class ClusterManagerId extends MapsObjectId<ClusterManager> {
+  /// Creates an immutable identifier for a [ClusterManager].
+  const ClusterManagerId(String value) : super(value);
+}
+
+/// TBD
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart:17]: Time to do this ;)

**Review by TimoPieti — COMMENTED** [review]
Might need some improvements and more documentation

File: packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart (hunk 1 above)
Comment (TimoPieti) [reply packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart:1172]: Yeah, which is scarier. The fact that the test passed or that there was a bug... :D

File: packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart (hunk 2 above)
Comment (TimoPieti) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart:118]: Oh, missed that.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java [hunk 7]
```diff
@@ -801,6 +850,19 @@ private void updateInitialMarkers() {
...
+  @Override
+  public void setInitialClusterManagers(Object initialClusterManagers) {
+    ArrayList<?> clusterManagers = (ArrayList<?>) initialClusterManagers;
+    this.initialClusterManagers = clusterManagers != null ? new ArrayList<>(clusterManagers) : null;
+    if (googleMap != null) {
+      updateInitialClusterManagers();
+    }
+  }
+
+  private void updateInitialClusterManagers() {
```
[the remark this answers is not in the document]
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:862]: This follows existing naming:
```java
    updateInitialClusterManagers();
    updateInitialMarkers();
    updateInitialPolygons();
    updateInitialPolylines();
    updateInitialCircles();
    updateInitialTileOverlays();
```

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    ClusterManager<MarkerBuilder> clusterManager =
+        new ClusterManager<>(context, googleMap, markerManager);
+    ClusterRenderer clusterRenderer = new ClusterRenderer(context, googleMap, clusterManager, this);
+    clusterManager.setRenderer(clusterRenderer);
+    initListenersForClusterManager(clusterManager, this, clusterMarkerListener);
+    clusterManagerIdToManager.put(clusterManagerId, clusterManager);
+  }
+
+  public void removeClusterManagers(List<Object> clusterManagerIdsToRemove) {
```
[the remark this answers is not in the document]
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:88]: Should.... I'll work on these

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart (hunk 6 above)
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart:17]: Oh, damn... yes!

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart [hunk 9]
```diff
@@ -174,6 +192,48 @@ class Marker implements MapsObject<Marker> {
     );
   }
 
+  /// Creates a new [Marker] object whose values are the same as this instance,
+  /// unless overwritten by the default values for specified parameters.
+  Marker copyWithDefaults({
```
Comment (TimoPieti) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart:197]: Do we need to test these on platform_interface lib tests?

**Review by TimoPieti — APPROVED** [review]
Check the last comment otherwise LGTM.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart (hunk 9 above)
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/marker.dart:197]: Test added, good catch

**Pull request merged**